### ISSUE TITLE
Redefining a function & Array

### ISSUE THREAD

**Issue by kalvotom** [JuliaLang/julia#21060]
I am experiencing a very confusing behavior (bug?) when redefining a function. A minimal example can be described in the following way:

```
julia> f(x) = x
f (generic function with 1 method)

julia> [f(x) for x in 1:3]
3-element Array{Int64,1}:
 1
 2
 3
```

Now redefine the function

```
julia> f(x) = 2*x
WARNING: Method definition f(Any) in module Main at REPL[2]:1 overwritten at REPL[4]:1.
f (generic function with 1 method)
```

and try it

```
julia> [f(1) for x in 1:3]
3-element Array{Int64,1}:
 2
 2
 2

julia> f(2)
4

julia> [f(x)+0 for x in 1:3]
3-element Array{Int64,1}:
 2
 4
 6
```

So far so good. BUT:

```
julia> [f(x) for x in 1:3]
3-element Array{Int64,1}:
 1
 2
 3

julia> [f(x) for x in 1:7]
7-element Array{Int64,1}:
 1
 2
 3
 4
 5
 6
 7
```

In these cases the old version of `f` is probably being called...

-----

Output of `versioninfo`:

```
julia> versioninfo()
Julia Version 0.5.1
Commit 6445c82* (2017-03-05 13:25 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i5-5200U CPU @ 2.20GHz
  WORD_SIZE: 64
  BLAS: libopenblas (DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas
  LIBM: libm
  LLVM: libLLVM-3.7.1 (ORCJIT, broadwell)
```

It looks like this bug does not affect 0.5.0 or 0.6.0-pre.alpha.52.

**Comment by yuyichao** [comment]
Dup of #265.

**Comment by kalvotom** [comment]
@yuyichao OK, I was not looking that far into the past since this behavior is not present in 0.5.0 and my version 0.5.1 is just a few days old. Sorry for duplication.

**Comment by tkelman** [comment]
> this behavior is not present in 0.5.0

Are you sure about that? There should be no difference in behavior between 0.5.0 and 0.5.1 on the code samples you posted.

**Comment by kalvotom** [comment]
@tkelman OK, I was sure. Now I checked it once more and I know I was wrong. What I thought was 0.5.0 was not actually 0.5.0 (but some dev version based on 0.5.0). Let me try this once more.

```
julia> versioninfo()
Julia Version 0.5.1
Commit 6445c82* (2017-03-05 13:25 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i5-5200U CPU @ 2.20GHz
  WORD_SIZE: 64
  BLAS: libopenblas (DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas
  LIBM: libm
  LLVM: libLLVM-3.7.1 (ORCJIT, broadwell)

julia> f(x) = x
f (generic function with 1 method)

julia> [f(x) for x in 1:3]
3-element Array{Int64,1}:
 1
 2
 3

julia> f(x) = 2*x
WARNING: Method definition f(Any) in module Main at REPL[2]:1 overwritten at REPL[4]:1.
f (generic function with 1 method)

julia> [f(x) for x in 1:3]
3-element Array{Int64,1}:
 1
 2
 3
```

and 

```
julia> versioninfo()                                                                                                                                                
Julia Version 0.5.0
Commit 3c9d753 (2016-09-19 18:14 UTC)
Platform Info:
  System: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU @ 2.30GHz
  WORD_SIZE: 64
  BLAS: libopenblas (NO_LAPACKE DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: liblapack.so.3
  LIBM: libopenlibm
  LLVM: libLLVM-3.7.1 (ORCJIT, haswell)

julia> f(x) = x                                                                                                                                                     
f (generic function with 1 method)
 
julia> [f(x) for x in 1:3]                                                                                                                                          
3-element Array{Int64,1}:
 1
 2
 3
 
julia> f(x) = 2*x                                                                                                                                                   
WARNING: Method definition f(Any) in module Main at REPL[1]:1 overwritten at REPL[3]:1.
f (generic function with 1 method)
 
julia> [f(x) for x in 1:3]                                                                                                                                          
3-element Array{Int64,1}:
 1
 2
 3
```

Indeed, 0.5.0 and 0.5.1 behaves in the same way.

**Comment by StefanKarpinski** [comment]
This is fixed in 0.6 – that's one of the major changes in 0.6.
